File: filecount.go
package main

import (
	"os"
	"time"
)

var g_filecount = 0

func iterate(path string) []string {
	var dirlist []string
	println(path)
	entries, err := os.ReadDir(path)
	if err != nil {
		panic(err)
	}
	for _, entry := range entries {
		if entry.IsDir() {
			dirlist = append(dirlist, path+"/"+entry.Name())
		} else {
			g_filecount++
		}
	}
	return dirlist
}

func countfiles(dir string) {
	dirlist := []string{dir}
	var directory string
	for {
		if len(dirlist) == 0 {
			break
		}
		directory, dirlist = dirlist[len(dirlist)-1], dirlist[:len(dirlist)-1]
		newdirs := iterate(directory)
		dirlist = append(dirlist, newdirs...)
	}
}

func main() {
	start := time.Now()
	countfiles("/tmp/data2")
	println(g_filecount)
	duration := time.Since(start)
	println(duration)
}

File: filecount.py
import os, time
from time import sleep
import shutil
from collections import deque
import os

filecount = 0

def iterate(path):
    global filecount
    dirlist = []
    with os.scandir(path) as it:
        for entry in it:
            if entry.is_dir():
                dirlist.append(entry)
            else:
                filecount += 1
    return dirlist


def countfiles(dir = '/tmp/data2'):
    dirlist = [dir]
    while True:
        if not dirlist:
            break
        directory = dirlist.pop()
        newdirs = iterate(directory)
        dirlist.extend(newdirs)
    
start = time.time()
countfiles()
print("total files:", filecount)
end = time.time()
print(end - start)
